<!DOCTYPE html>
<html lang="en">
<head>
        <title>8th Diary</title>
        <meta charset="utf-8" />
        <link rel="stylesheet" href="/theme/css/main.css" type="text/css" />
   

        <!--[if IE]>
                <script src="http://html5shiv.googlecode.com/svn/trunk/html5.js"></script><![endif]-->

        <!--[if lte IE 7]>
                <link rel="stylesheet" type="text/css" media="all" href="/css/ie.css"/>
                <script src="/js/IE8.js" type="text/javascript"></script><![endif]-->

        <!--[if lt IE 7]>
                <link rel="stylesheet" type="text/css" media="all" href="/css/ie6.css"/><![endif]-->

</head>

<body id="index" class="home">
        <header id="banner" class="body">
                <h1><a href="/index.html">Codon Alignment GSoC Homepage </a></h1>
                <nav><ul>
                                                                                                    <li class="active"><a href="/category/gsoc.html">GSoC</a></li>
                                    <li ><a href="/category/me.html">Me</a></li>
                                    <li ><a href="/category/timeline.html">Timeline</a></li>
                                </ul></nav>
        </header><!-- /#banner -->
        <section id="content" class="body">
  <article>
    <header>
      <h1 class="entry-title">
        <a href="8th-diary.html" rel="bookmark"
           title="Permalink to 8th Diary">8th Diary</a></h1>
          </header>

    <div class="entry-content">
      <footer class="post-info">
        <abbr class="published" title="2013-09-02T11:46:00">
                Mon 02 September 2013
        </abbr>

                <address class="vcard author">
                By <a class="url fn" href="/author/zheng-ruan.html">Zheng Ruan</a>
        </address>
        <p>In <a href="/category/gsoc.html">GSoC</a>. </p>
<p>tags: <a href="/tag/codon-alignment.html">Codon Alignment</a><a href="/tag/gsoc.html">GSoC</a></p></footer><!-- /.post-info -->      <div class="section" id="things-have-been-done">
<h2>Things have been done</h2>
<p>In the last two weeks, the major job I've done is YN00 method
support for dN/dS ratio estimation. This method is much more complex
than I expected and some typos in original paper (PMID: 10666704)
make my implementation even more difficult. The e.q. 9 of of the
paper is not the same with the yn00.c of PAML (line 803). In this
case, I have to go through yn00.c and make sure each step of my
implementation produces similar result with Yang's code. Now the
YN00 method in <tt class="docutils literal">cal_dn_ds</tt> is generally correct, except the
initial values for iteration entail to be modified. I also add
a <tt class="docutils literal">get_dn_ds_matrix()</tt> method to <tt class="docutils literal">CodonAlignment</tt> class so
that user will obtain a paiwise dN matrix and dS matrix from the
method. Another feature I've added is the multiple nucleotides
forward frameshift mutation support. An example of this can be
found below. However, due to the inherent difficult of using
regular expression to anchor protein with nucleotide, the more
than 2 nucleotides forward frameshift events should not occur at
the first or last anchor.</p>
</div>
<div class="section" id="some-examples">
<h2>Some Examples</h2>
<p>1) Now <tt class="docutils literal">cal_dn_ds()</tt> method is able to estimate dN/dS ratio using
NG86, LWL85, YN00 and ML methods.</p>
<div class="highlight"><pre><span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span><span class="p">,</span> <span class="n">AlignIO</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">CodonAlign</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio.Data.CodonTable</span> <span class="kn">import</span> <span class="n">generic_by_id</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">IUPAC</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio.CodonAlign.CodonSeq</span> <span class="kn">import</span> <span class="n">cal_dn_ds</span>

<span class="o">&gt;&gt;&gt;</span> <span class="n">aln</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&#39;pro.aln&#39;</span><span class="p">,</span> <span class="s">&#39;clustal&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">seq1</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&#39;rbcL_cucu_mRNA.fa&#39;</span><span class="p">,</span> <span class="s">&#39;fasta&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">IUPACUnambiguousDNA</span><span class="p">())</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">seq2</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&#39;rbcL_nic_mRNA.fa&#39;</span><span class="p">,</span> <span class="s">&#39;fasta&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">IUPACUnambiguousDNA</span><span class="p">())</span>

<span class="o">&gt;&gt;&gt;</span> <span class="n">codon_aln</span> <span class="o">=</span> <span class="n">CodonAlign</span><span class="o">.</span><span class="n">build</span><span class="p">(</span><span class="n">aln</span><span class="p">,</span> <span class="p">[</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">],</span> <span class="n">codon_table</span><span class="o">=</span><span class="n">generic_by_id</span><span class="p">[</span><span class="mi">11</span><span class="p">])</span>

<span class="o">&gt;&gt;&gt;</span> <span class="n">dN</span><span class="p">,</span> <span class="n">dS</span> <span class="o">=</span> <span class="n">cal_dn_ds</span><span class="p">(</span><span class="n">codon_aln</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">codon_aln</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">method</span><span class="o">=</span><span class="s">&#39;NG86&#39;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="k">print</span> <span class="n">dN</span><span class="p">,</span> <span class="n">dS</span><span class="p">,</span> <span class="n">dN</span><span class="o">/</span><span class="n">dS</span>
<span class="mf">0.0360453355368</span> <span class="mf">0.399965211638</span> <span class="mf">0.0901211767622</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">dN</span><span class="p">,</span> <span class="n">dS</span> <span class="o">=</span> <span class="n">cal_dn_ds</span><span class="p">(</span><span class="n">codon_aln</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">codon_aln</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">method</span><span class="o">=</span><span class="s">&#39;LWL85&#39;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="k">print</span> <span class="n">dN</span><span class="p">,</span> <span class="n">dS</span><span class="p">,</span> <span class="n">dN</span><span class="o">/</span><span class="n">dS</span>
<span class="mf">0.0376618724075</span> <span class="mf">0.357887756398</span> <span class="mf">0.105233754813</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">dN</span><span class="p">,</span> <span class="n">dS</span> <span class="o">=</span> <span class="n">cal_dn_ds</span><span class="p">(</span><span class="n">codon_aln</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">codon_aln</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">method</span><span class="o">=</span><span class="s">&#39;ML&#39;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="k">print</span> <span class="n">dN</span><span class="p">,</span> <span class="n">dS</span><span class="p">,</span> <span class="n">dN</span><span class="o">/</span><span class="n">dS</span>
<span class="mf">0.0342472820368</span> <span class="mf">0.420673537297</span> <span class="mf">0.0814105927765</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">dN</span><span class="p">,</span> <span class="n">dS</span> <span class="o">=</span> <span class="n">cal_dn_ds</span><span class="p">(</span><span class="n">codon_aln</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">codon_aln</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">method</span><span class="o">=</span><span class="s">&#39;YN00&#39;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="k">print</span> <span class="n">dN</span><span class="p">,</span> <span class="n">dS</span><span class="p">,</span> <span class="n">dN</span><span class="o">/</span><span class="n">dS</span>
<span class="mf">0.0327731426263</span> <span class="mf">0.525218485301</span> <span class="mf">0.0623990654241</span>
</pre></div>
<p>2) The following example will show how multiple nucleotides forward
frameshift events are supported:</p>
<div class="highlight"><pre><span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>

<span class="o">&gt;&gt;&gt;</span> <span class="n">seq1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&#39;ATGAAAAAGCACGAGTTACTTTGCCAAGGGACAAGTAACAAGCTCACCCAGTTGGGCACTTTTGAAGACCACTTTCTGAGCCTACAGAGGATGTTCAACAACTGTGAGGTGGTCGGGGGGGGGGGGGGCTTGGGAATTTGGAAATTACCTACATGCAGAGTAGTTACAACCTTTCTTTTCTCAAGACCATCCAGGAGGTTGCCGGCTATGTACTCATTGCCCTC&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">IUPACUnambiguousDNA</span><span class="p">()),</span> <span class="nb">id</span><span class="o">=</span><span class="s">&#39;pro1&#39;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">seq2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&#39;ATGAAAAAGCACGAGTTCTTTGCCAAGGGACAAGTAACAAGCTCACCCAGTTGGGCACTTTTGAAGACCACTTTCTGAGCCTACAGAGGATGTTCAACAATGTGAGGTGGTCCTTGGGAATTTGGAAATTACCTACATGCAGAGTAGTTACAACCTTTCTTTTCTCAAGACCATCCAGGAGGTTGCCGGCTATGTACTCATTGCCCTC&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">IUPACUnambiguousDNA</span><span class="p">()),</span> <span class="nb">id</span><span class="o">=</span><span class="s">&#39;pro2&#39;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">seq3</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&#39;ATGAAAAAGCACGAGTTACTTTGCCAAGGGACAAGTAACAAGCTCACCCTTGGGCACTTTTGAAGACCACTTTCTGAGCCTACAGAGGATGTTCAACAACTGTGAGGTGGTCCTTGGGAATTTGGAAATTACCTACATGCAGAGTAGTTACAACCTTTCTTTTCTCAAGACCATCCAGGAGGTTGCCGGCTATGTACTCATTGCCCTC&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">IUPACUnambiguousDNA</span><span class="p">()),</span> <span class="nb">id</span><span class="o">=</span><span class="s">&#39;pro3&#39;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">seq4</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&#39;ATGAAAAAGCACGAGTTACTTTGCCAAGGGACAAGTAACAAGCTCACCCAGTTGGGCACTTTTGAAGACCACTTTCTGAGCCTACAGAGGATGTTCAACAACTGTGAGGTGGTCCTTGGGAATTTGGAAATTACCTACATGCAGAGTAGTTACAACCTTTCTTTTCTCAAGACCATCCAGGAGGTTGCCGGCTATGTACTCATTGCCCTC&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">IUPACUnambiguousDNA</span><span class="p">()),</span> <span class="nb">id</span><span class="o">=</span><span class="s">&#39;pro4&#39;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">seq5</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&#39;ATGAAAAAGCACGAGTTACTTTGCCAAGGGACAAGTAACAAGCTCACCCAGTTGGGCACTTTTGGGAAGACCACTTTCTGAGCCTACAGAGGATGTTCAACAACTGTGAGGTGGTCCTTGGGAATTTGGAAATTACCTACATGCAGAGTAGTTACAACCTTTCTTTTCTCAAGACCATCCAGGAGGTTGCCGGCTATGTACTCATTGCCCTC&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">IUPACUnambiguousDNA</span><span class="p">()),</span> <span class="nb">id</span><span class="o">=</span><span class="s">&#39;pro5&#39;</span><span class="p">)</span>

<span class="o">&gt;&gt;&gt;</span> <span class="n">pro1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&#39;MKKHELLCQGTSNKLTQLGTFEDHFLSLQRMFNNCEVVLGNLEITYMQSSYNLSFLKTIQEVAGYVLIAL&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s">&#39;pro1&#39;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">pro2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&#39;MKKHEFLCQGTSNKLTQLGTFEDHFLSLQRMFNNCEVVLGNLEITYMQSSYNLSFLKTIQEVAGYVLIAL&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s">&#39;pro2&#39;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">pro3</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&#39;MKKHELLCQGTSNKLTLLGTFEDHFLSLQRMFNNCEVVLGNLEITYMQSSYNLSFLKTIQEVAGYVLIAL&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s">&#39;pro3&#39;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">pro4</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&#39;MKKHELLCQGTSNKLTQLGTFEDHFLSLQRMFNNCEVVLGNLEITYMCSSYNLSFLKTIQEVAGYVLIAL&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s">&#39;pro4&#39;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">pro5</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&#39;MKKHELLCQGTSNKLTQLGTFEDHFLSLQRMFNNCEVVLGNLEITYMQSSYNLSFLKTIQEVAGYVLIAL&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s">&#39;pro5&#39;</span><span class="p">)</span>

<span class="o">&gt;&gt;&gt;</span> <span class="n">aln</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span><span class="n">pro1</span><span class="p">,</span> <span class="n">pro2</span><span class="p">,</span> <span class="n">pro3</span><span class="p">,</span> <span class="n">pro4</span><span class="p">,</span> <span class="n">pro5</span><span class="p">])</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">codon_aln</span> <span class="o">=</span> <span class="n">build</span><span class="p">(</span><span class="n">aln</span><span class="p">,</span> <span class="p">[</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span> <span class="n">seq3</span><span class="p">,</span> <span class="n">seq4</span><span class="p">,</span> <span class="n">seq5</span><span class="p">],</span> <span class="n">anchor_len</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>

<span class="o">&gt;&gt;&gt;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">codon_aln</span><span class="p">:</span>
<span class="o">...</span>     <span class="k">print</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">rz</span><span class="o">/</span><span class="n">Working</span><span class="o">.</span><span class="n">Dir</span><span class="o">/</span><span class="n">GSoC</span><span class="o">/</span><span class="n">biopython</span><span class="o">/</span><span class="n">Bio</span><span class="o">/</span><span class="n">CodonAlign</span><span class="o">/</span><span class="n">__init__</span><span class="o">.</span><span class="n">py</span><span class="p">:</span><span class="mi">578</span><span class="p">:</span> <span class="ne">UserWarning</span><span class="p">:</span> <span class="n">pro4</span><span class="p">(</span><span class="n">C</span> <span class="mi">47</span><span class="p">)</span> <span class="n">does</span> <span class="ow">not</span> <span class="n">correspond</span> <span class="n">to</span> <span class="n">pro4</span><span class="p">(</span><span class="n">CAG</span><span class="p">)</span>
  <span class="o">%</span> <span class="p">(</span><span class="n">pro</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">aa</span><span class="p">,</span> <span class="n">aa_num</span><span class="p">,</span> <span class="n">nucl</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">this_codon</span><span class="p">))</span>
  <span class="n">ATGAAAAAGCACGAGTTACTTTGCCAAGGGACAAGTAACAAGCTCACCCAGTTGGGCACTTTT</span><span class="o">---</span><span class="n">GAAGACCACTTTCTGAGCCTACAGAGGATGTTCAACAACTGTGAGGTGGTC</span><span class="o">-</span><span class="n">GGGGGGGGGGGGGGCTTGGGAATTTGGAAATTACCTACATGCAGAGTAGTTACAACCTTTCTTTTCTCAAGACCATCCAGGAGGTTGCCGGCTATGTACTCATTGCCCTC</span>
  <span class="n">ATGAAAAAGCACGAGTT</span><span class="o">-</span><span class="n">CTTTGCCAAGGGACAAGTAACAAGCTCACCCAGTTGGGCACTTTT</span><span class="o">---</span><span class="n">GAAGACCACTTTCTGAGCCTACAGAGGATGTTCAACAA</span><span class="o">-</span><span class="n">TGTGAGGTGGTC</span><span class="o">---------------</span><span class="n">CTTGGGAATTTGGAAATTACCTACATGCAGAGTAGTTACAACCTTTCTTTTCTCAAGACCATCCAGGAGGTTGCCGGCTATGTACTCATTGCCCTC</span>
  <span class="n">ATGAAAAAGCACGAGTTACTTTGCCAAGGGACAAGTAACAAGCTCACCC</span><span class="o">--</span><span class="n">TTGGGCACTTTT</span><span class="o">---</span><span class="n">GAAGACCACTTTCTGAGCCTACAGAGGATGTTCAACAACTGTGAGGTGGTC</span><span class="o">---------------</span><span class="n">CTTGGGAATTTGGAAATTACCTACATGCAGAGTAGTTACAACCTTTCTTTTCTCAAGACCATCCAGGAGGTTGCCGGCTATGTACTCATTGCCCTC</span>
  <span class="n">ATGAAAAAGCACGAGTTACTTTGCCAAGGGACAAGTAACAAGCTCACCCAGTTGGGCACTTTT</span><span class="o">---</span><span class="n">GAAGACCACTTTCTGAGCCTACAGAGGATGTTCAACAACTGTGAGGTGGTC</span><span class="o">---------------</span><span class="n">CTTGGGAATTTGGAAATTACCTACATGCAGAGTAGTTACAACCTTTCTTTTCTCAAGACCATCCAGGAGGTTGCCGGCTATGTACTCATTGCCCTC</span>
  <span class="n">ATGAAAAAGCACGAGTTACTTTGCCAAGGGACAAGTAACAAGCTCACCCAGTTGGGCACTTTTGG</span><span class="o">-</span><span class="n">GAAGACCACTTTCTGAGCCTACAGAGGATGTTCAACAACTGTGAGGTGGTC</span><span class="o">---------------</span><span class="n">CTTGGGAATTTGGAAATTACCTACATGCAGAGTAGTTACAACCTTTCTTTTCTCAAGACCATCCAGGAGGTTGCCGGCTATGTACTCATTGCCCTC</span>
</pre></div>
<p>From the alignment above, we can see multiple Gs are shifted and gaps are added to the corresponding region of other records.</p>
</div>
<div class="section" id="future-plan">
<h2>Future Plan</h2>
<ol class="arabic simple">
<li>Add more options for ML method (allow different substitution model);</li>
<li>Implement mcdonald kreitman neutrality test;</li>
<li>Tree estimation using dN, dS distance;</li>
<li>Documentation.</li>
</ol>
<p>For more info, please visit my
[CodonAlign github repo](<a class="reference external" href="https://github.com/zruan/biopython/tree/master/Bio/CodonAlign">https://github.com/zruan/biopython/tree/master/Bio/CodonAlign</a>)</p>
</div>

    </div><!-- /.entry-content -->
        <div class="comments">
      <h2>Comments !</h2>
      <div id="disqus_thread"></div>
      <script type="text/javascript">
        var disqus_identifier = "8th-diary.html";
        (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = 'http://zruansblog.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
      </script>
    </div>
    
  </article>
</section>
        <section id="extras" class="body">
                        <div class="blogroll">
                        <h2>blogroll</h2>
                        <ul>
                                                    <li><a href="http://informatics.nescent.org/wiki/Phyloinformatics_Summer_of_Code_2013#Codon_Alignment_and_Analysis_in_Biopython">NESCent GSoC Project wiki</a></li>
                                                    <li><a href="http://biopython.org/wiki/Main_Page">Biopython</a></li>
                                                    <li><a href="http://iob.uga.edu/">UGA Bioinformatics</a></li>
                                                    <li><a href="http://getpelican.com/">Pelican</a></li>
                                                    <li><a href="http://python.org/">Python.org</a></li>
                                                    <li><a href="http://jinja.pocoo.org/">Jinja2</a></li>
                                                </ul>
                </div><!-- /.blogroll -->
                                <div class="social">
                        <h2>social</h2>
                        <ul>
                            
                                                    <li><a href="https://www.facebook.com/ruan.zheng.7">Facebook</a></li>
                                                    <li><a href="http://zr1991.blogspot.com/">Google Blogger</a></li>
                                                    <li><a href="http://user.qzone.qq.com/8390905">Qzone</a></li>
                                                </ul>
                </div><!-- /.social -->
                </section><!-- /#extras -->

        <footer id="contentinfo" class="body">

        </footer><!-- /#contentinfo -->

<script type="text/javascript">
    var disqus_shortname = 'zruansblog';
    (function () {
        var s = document.createElement('script'); s.async = true;
        s.type = 'text/javascript';
        s.src = 'http://' + disqus_shortname + '.disqus.com/count.js';
        (document.getElementsByTagName('HEAD')[0] || document.getElementsByTagName('BODY')[0]).appendChild(s);
    }());
</script>
</body>
</html>