<!DOCTYPE html>
<html lang="en">
<head>
        <title>Codon Alignment</title>
        <meta charset="utf-8" />
                <link rel="stylesheet" href="/theme/css/main.css" type="text/css" />
        <link href="/" type="application/atom+xml" rel="alternate" title="Codon Alignment GSoC Homepage ATOM Feed" />
        

        <!--[if IE]>
                <script src="http://html5shiv.googlecode.com/svn/trunk/html5.js"></script><![endif]-->

        <!--[if lte IE 7]>
                <link rel="stylesheet" type="text/css" media="all" href="/css/ie.css"/>
                <script src="/js/IE8.js" type="text/javascript"></script><![endif]-->

        <!--[if lt IE 7]>
                <link rel="stylesheet" type="text/css" media="all" href="/css/ie6.css"/><![endif]-->

</head>

<body id="index" class="home">
        <header id="banner" class="body">
                <h1><a href="">Codon Alignment GSoC Homepage </a></h1>
                <nav><ul>
                                                                                        <li ><a href="/category/gsoc.html">GSoC</a></li>
                                    <li ><a href="/category/me.html">Me</a></li>
                                                </ul></nav>
        </header><!-- /#banner -->
                
<section id="content" class="body">    
<article>
        <header> <h1 class="entry-title"><a href="codon-alignment.html"
        rel="bookmark" title="Permalink to Codon Alignment">Codon Alignment</a></h1>   </header>
        <div class="entry-content">
        <footer class="post-info">
        <abbr class="published" title="2013-06-04T23:53:00">
                Tue 04 June 2013
        </abbr>

                <address class="vcard author">
                By <a class="url fn" href="/author/zheng-ruan.html">Zheng Ruan</a>
        </address>
        <p>In <a href="/category/gsoc.html">GSoC</a>. </p>
<p>tags: <a href="/tag/codon-alignment.html">Codon Alignment</a><a href="/tag/gsoc.html">GSoC</a></p></p>
</footer><!-- /.post-info -->        <h1>Abstract</h1>
<p>A codon alignment is an alignment of nucleotide sequences in which 
trinucleotides correspond directly to amino acids in translated 
product. Codon alignment is useful as it distinguished different 
types of nucleotides substitution and is frequently used to test 
neutrality and calculate selection pressure. My job in the project
is to develop codon alignment support for biopython. This will 
facilitate quick and large scale evolutionary analysis in biopython
environment.</p>
<h1>Project Expectation</h1>
<ol>
<li>Generate codon alignment based on protein alignment and their
corresponding nucleotide sequences. The functionality is able to
deal with conditions like mismatches, frameshift, extra sequences 
in 5'UTR and 3'UTR, as implemented in pal2nal;</li>
<li>Estimate nonsynonymous sites and synonymous sites using counting
based method, such as NG86 [1], LPB93 [2], LWL85 [3], LWL85m [4],
Ina93 [5], YN00 [6], physical-site distance [7].</li>
<li>Estimate nonsynonymous sites and synonymous sites based on ML
method;</li>
<li>McDonald-Kreitman method to test neutrality; Estimate selection
for each codon (HyPhy); Codon based Z-test and Fisher’s Exact test
of selection. </li>
<li>Identify positive selection sites using dn/ds ratio and Naive
Empirical Bayesian approach.</li>
<li>Other miscellaneous useful function.</li>
</ol>
<h1>References</h1>
<ul>
<li>[1] Nei M and Gojobori T. (1986). Simple methods for estimating 
the numbers of synonymous and nonsynonymous nucleotide
substitutions. <em>Mol. Biol. Evol</em> 3:418-426.</li>
<li>[2] Li WH. (1993). Unbiased estimation of the rates of synonymous 
and nonsynonymous substitution. <em>J Mol Evol.</em> 36(1):96-99.</li>
<li>[3] Li, W.-H., C.-I. Wu, and C.-C. Luo. (1985). A new method for
estimating synonymous and non-synonymous rates of nucleotide
substitutions considering the relative likelihood of nucleotide
and codon changes. <em>Mol. Biol. Evol.</em> 2:150–174.</li>
<li>[4] Yun-Huei Tzeng, Runsun Pan, and Wen-Hsiung Li. (2004).
Comparison of Three Methods for Estimating Rates of Synonymous
and Nonsynonymous Nucleotide Substitutions. <em>Mol. Bio Evol.</em>
21(12):2290-2298.</li>
<li>[5] Ina, Y. (1995). New methods for estimating the numbers of
synonymous and nonsynonymous substitutions. <em>J. Mol. Evol.</em>
40:190–226.</li>
<li>[6] Yang Z, Nielsen R. (2000). Estimating synonymous and
nonsynonymous substitution rates under realistic evolutionary
models. <em>Mol Biol Evol.</em> 17:32-43.</li>
<li>[7] Yang Z. (2006). Computational molecular evolution. Oxford
(UK): Oxford University Press.</li>
</ul>
<p>For more info, please visit my
<a href="http://github.com/zruan/CodonAlign">codonAlign github repo</a></p>
        </div><!-- /.entry-content -->
                <div class="comments">
        <h2>Comments !</h2>
            <div id="disqus_thread"></div>
            <script type="text/javascript">
               var disqus_identifier = "codon-alignment.html";
               (function() {
               var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
               dsq.src = 'http://zruansblog.disqus.com/embed.js';
               (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
              })();
            </script>
        </div>
        
</article>
</section>
        <section id="extras" class="body">
                        <div class="blogroll">
                        <h2>blogroll</h2>
                        <ul>
                                                    <li><a href="http://informatics.nescent.org/wiki/Phyloinformatics_Summer_of_Code_2013#Codon_Alignment_and_Analysis_in_Biopython">NESCent GSoC Project wiki</a></li>
                                                    <li><a href="http://biopython.org/wiki/Main_Page">Biopython</a></li>
                                                    <li><a href="http://iob.uga.edu/">UGA Bioinformatics</a></li>
                                                    <li><a href="http://getpelican.com/">Pelican</a></li>
                                                    <li><a href="http://python.org/">Python.org</a></li>
                                                    <li><a href="http://jinja.pocoo.org/">Jinja2</a></li>
                                                </ul>
                </div><!-- /.blogroll -->
                                <div class="social">
                        <h2>social</h2>
                        <ul>
                            <li><a href="/" rel="alternate">atom feed</a></li>
                            
                                                    <li><a href="https://www.facebook.com/ruan.zheng.7">Facebook</a></li>
                                                    <li><a href="http://zr1991.blogspot.com/">Google Blogger</a></li>
                                                    <li><a href="http://user.qzone.qq.com/8390905">Qzone</a></li>
                                                </ul>
                </div><!-- /.social -->
                </section><!-- /#extras -->

        <footer id="contentinfo" class="body">
                <address id="about" class="vcard body">
                Proudly powered by <a href="http://alexis.notmyidea.org/pelican/">pelican</a>, which takes great advantages of <a href="http://python.org">python</a>.
                </address><!-- /#about -->

                <p>The theme is by <a href="http://smashingmagazine.com">Smashing Magazine</a>, thanks!</p>
        </footer><!-- /#contentinfo -->

<script type="text/javascript">
    var disqus_shortname = 'zruansblog';
    (function () {
        var s = document.createElement('script'); s.async = true;
        s.type = 'text/javascript';
        s.src = 'http://' + disqus_shortname + '.disqus.com/count.js';
        (document.getElementsByTagName('HEAD')[0] || document.getElementsByTagName('BODY')[0]).appendChild(s);
    }());
</script>
</body>
</html>