<!DOCTYPE html>
<html lang="en">
<head>
        <title>Codon Alignment GSoC Homepage - Codon Alignment</title>
        <meta charset="utf-8" />
        <link rel="stylesheet" href="/theme/css/main.css" type="text/css" />
   

        <!--[if IE]>
                <script src="http://html5shiv.googlecode.com/svn/trunk/html5.js"></script><![endif]-->

        <!--[if lte IE 7]>
                <link rel="stylesheet" type="text/css" media="all" href="/css/ie.css"/>
                <script src="/js/IE8.js" type="text/javascript"></script><![endif]-->

        <!--[if lt IE 7]>
                <link rel="stylesheet" type="text/css" media="all" href="/css/ie6.css"/><![endif]-->

</head>

<body id="index" class="home">
        <header id="banner" class="body">
                <h1><a href="/index.html">Codon Alignment GSoC Homepage </a></h1>
                <nav><ul>
                                                                                                    <li ><a href="/category/gsoc.html">GSoC</a></li>
                                    <li ><a href="/category/me.html">Me</a></li>
                                </ul></nav>
        </header><!-- /#banner -->
                
            

                            <aside id="featured" class="body">
                <article>
                    <h1 class="entry-title"><a href="/1st-diary.html">1st Diary</a></h1> 
                    <footer class="post-info">
        <abbr class="published" title="2013-06-20T14:32:00">
                Thu 20 June 2013
        </abbr>

                <address class="vcard author">
                By <a class="url fn" href="/author/zheng-ruan.html">Zheng Ruan</a>
        </address>
        <p>In <a href="/category/gsoc.html">GSoC</a>. </p>
<p>tags: <a href="/tag/codon-alignment.html">Codon Alignment</a><a href="/tag/gsoc.html">GSoC</a></p></footer><!-- /.post-info --><div class="section" id="things-have-been-done">
<h2>Things have been done</h2>
<p>In the first 4 days of GSoC, I successfully ported my existing
code into a forked biopython distribution in my
<a class="reference external" href="https://github.com/zruan/biopython/tree/master/Bio/CodonAlign">github</a>.
I didn't pay much effort to the code organization as all of them
are put into <tt class="docutils literal">__init__.py</tt>. Maybe when the code grows larger I
need to make them more organized. The major class for <tt class="docutils literal">CodonAlign</tt>
is <tt class="docutils literal">CodonAlignment</tt> that inherits from <tt class="docutils literal">MultipleSeqAlignment</tt>.
I also defined a <tt class="docutils literal">CodonSeq</tt> class (from <tt class="docutils literal">Seq</tt>) to store sequence
that is aligned based on codons. The main difference is that the <tt class="docutils literal">__getitem__()</tt>
method of <tt class="docutils literal">CodonSeq</tt> is codon slice (i.e. three nucleotides per slice,
see below example). I also changed the <tt class="docutils literal">__str__()</tt>
method so that it is more suitable to print a <tt class="docutils literal">CodonAlignment</tt> into the screen.
Another thing I've done is the <tt class="docutils literal">build</tt> function, which accept a
protein <tt class="docutils literal">MultipleSeqAlignment</tt> object and a set of corresponding
nucleotide <tt class="docutils literal">SeqRecord</tt> to come up with a
<tt class="docutils literal">CodonAlignment</tt>. Internally, the function will translate the protein
sequences into a long regular expressions based on the user
specified <tt class="docutils literal">CodonTable</tt> in biopython. It will then use the regualr
expression to scan the nucleotide sequence to find a span.
No mismatch and frameshift is allowed right now in this step.
Finally, a <tt class="docutils literal">CodonAlignment</tt> will be returned to the user for
further manipulation.</p>
</div>
<div class="section" id="some-examples">
<h2>Some Examples</h2>
<p>It is best to illustrate <tt class="docutils literal">CodonAlignment</tt> using simple examples.
Here are some (assuming <tt class="docutils literal">__init__.py</tt> and <tt class="docutils literal">test_data</tt> are in your
current python directory):</p>
<p>1. Build a <tt class="docutils literal">CodonAlignment</tt> from Protein Alignment and Nucleotide
Sequences. In the following example, <tt class="docutils literal">nucl_seq1</tt> is an iterator
returned by <tt class="docutils literal">SeqIO.parse()</tt>. I assume the order of sequence
records stored in <tt class="docutils literal">prot_aln1</tt> and <tt class="docutils literal">nucl_seq1</tt> are the same.</p>
<div class="highlight"><pre><span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span><span class="p">,</span> <span class="n">AlignIO</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">IUPAC</span>
<span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">__init__</span> <span class="kn">import</span> <span class="n">build</span><span class="p">,</span> <span class="n">default_alphabet</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">prot_aln1</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&#39;test_data/pro1.aln&#39;</span><span class="p">,</span> <span class="s">&#39;clustal&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">nucl_seq1</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s">&#39;test_data/nucl1.fa&#39;</span><span class="p">,</span> <span class="s">&#39;fasta&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">default_alphabet</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">codon_aln1</span> <span class="o">=</span> <span class="n">build</span><span class="p">(</span><span class="n">prot_aln1</span><span class="p">,</span> <span class="n">nucl_seq1</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="k">print</span> <span class="n">codon_aln1</span>
<span class="n">CodonAlignment</span> <span class="n">Object</span>
<span class="n">HasStopCodon</span><span class="p">(</span><span class="n">Gapped</span><span class="p">(</span><span class="n">IUPACUnambiguousDNA</span><span class="p">(),</span> <span class="s">&#39;-&#39;</span><span class="p">),</span> <span class="s">&#39;*&#39;</span><span class="p">)</span> <span class="n">alignment</span> <span class="k">with</span> <span class="mi">3</span> <span class="n">rows</span> <span class="ow">and</span> <span class="mi">555</span> <span class="n">columns</span> <span class="p">(</span><span class="mi">185</span> <span class="n">codons</span><span class="p">)</span>
<span class="n">AGAGGCGATCAACGCAGCAACTTCCAGCTGTCTCCCTCCACCATGCAGATC</span><span class="o">...</span><span class="n">GGGGCATGC</span> <span class="n">isotig69710</span>
<span class="o">------------------------------------------</span><span class="n">ATGAAAGTC</span><span class="o">...---------</span> <span class="n">ENSG00000108691</span><span class="p">:</span><span class="n">ENST00000580907</span>
<span class="o">------------------------------------------</span><span class="n">ATGAAAGTC</span><span class="o">...---------</span> <span class="n">ENSG00000108691</span><span class="p">:</span><span class="n">ENST00000225831</span>
</pre></div>
<p>2. It is also possible to use SeqRecord dict read by
<tt class="docutils literal">SeqIO.index()</tt> to build <tt class="docutils literal">CodonAlignment</tt>. In this case, I assume
nucleotide <tt class="docutils literal">SeqRecord</tt> and protein <tt class="docutils literal">SeqRecord</tt> sharing the same
id are a match. e.g.</p>
<div class="highlight"><pre><span class="o">&gt;&gt;&gt;</span> <span class="n">prot_aln2</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&#39;test_data/pro3.aln&#39;</span><span class="p">,</span> <span class="s">&#39;clustal&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">nucl_seq2</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s">&#39;test_data/nucl3.fa&#39;</span><span class="p">,</span> <span class="s">&#39;fasta&#39;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">default_alphabet</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">codon_aln2</span> <span class="o">=</span> <span class="n">build</span><span class="p">(</span><span class="n">prot_aln2</span><span class="p">,</span> <span class="n">nucl_seq2</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="k">print</span> <span class="n">codon_aln2</span>
<span class="n">CodonAlignment</span> <span class="n">Object</span>
<span class="n">HasStopCodon</span><span class="p">(</span><span class="n">Gapped</span><span class="p">(</span><span class="n">IUPACUnambiguousDNA</span><span class="p">(),</span> <span class="s">&#39;-&#39;</span><span class="p">),</span> <span class="s">&#39;*&#39;</span><span class="p">)</span> <span class="n">alignment</span> <span class="k">with</span> <span class="mi">10</span> <span class="n">rows</span> <span class="ow">and</span> <span class="mi">3366</span> <span class="n">columns</span> <span class="p">(</span><span class="mi">1122</span> <span class="n">codons</span><span class="p">)</span>
<span class="o">---------------------------------------------------...</span><span class="n">CCCAGTTGG</span> <span class="n">ENSG00000166888</span><span class="p">:</span><span class="n">ENST00000300134</span>
<span class="o">---------------------------------------------------...</span><span class="n">CCCAGTTGG</span> <span class="n">ENSG00000166888</span><span class="p">:</span><span class="n">ENST00000543873</span>
<span class="o">---------------------------------------------------...</span><span class="n">CCCAGTTGG</span> <span class="n">ENSG00000166888</span><span class="p">:</span><span class="n">ENST00000556155</span>
<span class="o">---------------------------------------------------...</span><span class="n">CCCAGTTGG</span> <span class="n">ENSG00000166888</span><span class="p">:</span><span class="n">ENST00000454075</span>
<span class="o">---------------------------------------------------...</span><span class="n">CCCAGTTGG</span> <span class="n">ENSG00000166888</span><span class="p">:</span><span class="n">ENST00000538913</span>
<span class="o">---------------------------------------------------...</span><span class="n">CCCAGTTGG</span> <span class="n">ENSG00000166888</span><span class="p">:</span><span class="n">ENST00000537215</span>
<span class="o">---------------------------------------------------...</span><span class="n">CCCAGTTGG</span> <span class="n">ENSG00000166888</span><span class="p">:</span><span class="n">ENST00000535201</span>
<span class="n">GAGGTTACTCAGAGTAGGAGGAAGCCGGTCAGAGAGGGCAGACCCTGGGAA</span><span class="o">...</span><span class="n">CTGCTTTGC</span> <span class="n">isotig46679</span>
<span class="o">------------</span><span class="n">GCTAGGAGA</span><span class="o">------------------------------...---------</span> <span class="n">isotig12565</span>
<span class="o">------------</span><span class="n">GCTAGGAGA</span><span class="o">------------------------------...---------</span> <span class="n">isotig12566</span>
</pre></div>
<p>3. Then let's look at the <tt class="docutils literal">SeqRecord</tt> in <tt class="docutils literal">CodonAlignment</tt>. The
<tt class="docutils literal">CodonSeq</tt> object stored in it is a bit different from normal <tt class="docutils literal">Seq</tt>
Object. The main difference it that it is sliced by codon not
nucleotide. Let's see the following example.</p>
<div class="highlight"><pre><span class="o">&gt;&gt;&gt;</span> <span class="n">codon_record</span> <span class="o">=</span> <span class="n">codon_aln1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="o">&gt;&gt;&gt;</span> <span class="k">print</span> <span class="n">codon_record</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">seq</span><span class="o">=</span><span class="n">CodonSeq</span><span class="p">(</span><span class="s">&#39;AGAGGCGATCAACGCAGCAACTTCCAGCTGTCTCCCTCCACCATGCAGATCTCC...TGC&#39;</span><span class="p">,</span> <span class="n">Gapped</span><span class="p">(</span><span class="n">HasStopCodon</span><span class="p">(</span><span class="n">IUPACUnambiguousDNA</span><span class="p">(),</span> <span class="s">&#39;*&#39;</span><span class="p">),</span> <span class="s">&#39;-&#39;</span><span class="p">)),</span> <span class="nb">id</span><span class="o">=</span><span class="s">&#39;isotig69710&#39;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s">&#39;&lt;unknown name&gt;&#39;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s">&#39;&lt;unknown description&gt;&#39;</span><span class="p">,</span> <span class="n">dbxrefs</span><span class="o">=</span><span class="p">[])</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">codon_record</span><span class="o">.</span><span class="n">seq</span>
<span class="n">CodonSeq</span><span class="p">(</span><span class="s">&#39;AGAGGCGATCAACGCAGCAACTTCCAGCTGTCTCCCTCCACCATGCAGATCTCC...TGC&#39;</span><span class="p">,</span> <span class="n">Gapped</span><span class="p">(</span><span class="n">HasStopCodon</span><span class="p">(</span><span class="n">IUPACUnambiguousDNA</span><span class="p">(),</span> <span class="s">&#39;*&#39;</span><span class="p">),</span> <span class="s">&#39;-&#39;</span><span class="p">))</span>
<span class="o">&gt;&gt;&gt;</span> <span class="k">print</span> <span class="n">codon_record</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">__class__</span><span class="o">.</span><span class="n">__name__</span>
<span class="n">CodonSeq</span>
<span class="o">&gt;&gt;&gt;</span> <span class="k">print</span> <span class="n">codon_record</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">AGA</span>
<span class="o">&gt;&gt;&gt;</span> <span class="k">print</span> <span class="n">codon_record</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span>
<span class="n">AGAGGCGAT</span>
<span class="o">&gt;&gt;&gt;</span> <span class="k">print</span> <span class="n">codon_record</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="o">-</span><span class="mi">3</span><span class="p">:]</span>
<span class="n">GGGGCATGC</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">codon_record</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">get_codon_num</span><span class="p">()</span>
<span class="mi">185</span>
</pre></div>
<p>The examples above are the basic usage of <tt class="docutils literal">CodonAlignment</tt>. If you
have any suggestions or request some features that are useful for codon
alignment please let me know.</p>
</div>
<div class="section" id="future-plan">
<h2>Future Plan</h2>
<p>1. Automatically generate an alphabet from user specified
<tt class="docutils literal">CodonTable</tt>. And check the <tt class="docutils literal">CodonSeq</tt> to see if the alphabet in the
sequence is compatible.</p>
<p>For more info, please visit my
[codonAlign github repo](<a class="reference external" href="https://github.com/zruan/biopython/tree/master/Bio/CodonAlign">https://github.com/zruan/biopython/tree/master/Bio/CodonAlign</a>)</p>
</div>
<p>There are <a href="/1st-diary.html#disqus_thread">comments</a>.</p>                </article>
                            </aside><!-- /#featured -->
                            <section id="content" class="body">
                    <h1>Other articles</h1>
                    <hr />
                    <ol id="posts-list" class="hfeed">
                                                        

                 
                        <li><article class="hentry">    
                <header>
                        <h1><a href="/project-timeline.html" rel="bookmark" title="Permalink to Project Timeline">Project Timeline</a></h1>
                </header>
                
                <div class="entry-content">
                <footer class="post-info">
        <abbr class="published" title="2013-06-14T15:24:00">
                Fri 14 June 2013
        </abbr>

                <address class="vcard author">
                By <a class="url fn" href="/author/zheng-ruan.html">Zheng Ruan</a>
        </address>
        <p>In <a href="/category/gsoc.html">GSoC</a>. </p>
<p>tags: <a href="/tag/codon-alignment.html">Codon Alignment</a><a href="/tag/gsoc.html">GSoC</a><a href="/tag/timeline.html">Timeline</a></p></footer><!-- /.post-info -->                <p>This timeline is set up according to Karen Cranston's suggestion.
I will update this page according to the progress of my current
coding status and suggestions from others. My original proposal also contains a draft
timeline, which I will still generally follow. It can be found at
<a href="https://google-melange.appspot.com/gsoc/proposal/review/google/gsoc2013/rzzmh12345/1">Melange</a>.
Hope ...</p>
                <a class="readmore" href="/project-timeline.html">read more</a>
                <p>There are <a href="/project-timeline.html#disqus_thread">comments</a>.</p>                </div><!-- /.entry-content -->
            </article></li>
                                    

                 
                        <li><article class="hentry">    
                <header>
                        <h1><a href="/codon-alignment.html" rel="bookmark" title="Permalink to Codon Alignment">Codon Alignment</a></h1>
                </header>
                
                <div class="entry-content">
                <footer class="post-info">
        <abbr class="published" title="2013-06-04T23:53:00">
                Tue 04 June 2013
        </abbr>

                <address class="vcard author">
                By <a class="url fn" href="/author/zheng-ruan.html">Zheng Ruan</a>
        </address>
        <p>In <a href="/category/gsoc.html">GSoC</a>. </p>
<p>tags: <a href="/tag/codon-alignment.html">Codon Alignment</a><a href="/tag/gsoc.html">GSoC</a></p></footer><!-- /.post-info -->                <p>General Description of Codon Alignment Project</p>
                <a class="readmore" href="/codon-alignment.html">read more</a>
                <p>There are <a href="/codon-alignment.html#disqus_thread">comments</a>.</p>                </div><!-- /.entry-content -->
            </article></li>
                            <p class="paginator">
        Page 1 / 1
    </p>
                            </ol><!-- /#posts-list -->
            </section><!-- /#content -->
                    <section id="extras" class="body">
                        <div class="blogroll">
                        <h2>blogroll</h2>
                        <ul>
                                                    <li><a href="http://informatics.nescent.org/wiki/Phyloinformatics_Summer_of_Code_2013#Codon_Alignment_and_Analysis_in_Biopython">NESCent GSoC Project wiki</a></li>
                                                    <li><a href="http://biopython.org/wiki/Main_Page">Biopython</a></li>
                                                    <li><a href="http://iob.uga.edu/">UGA Bioinformatics</a></li>
                                                    <li><a href="http://getpelican.com/">Pelican</a></li>
                                                    <li><a href="http://python.org/">Python.org</a></li>
                                                    <li><a href="http://jinja.pocoo.org/">Jinja2</a></li>
                                                </ul>
                </div><!-- /.blogroll -->
                                <div class="social">
                        <h2>social</h2>
                        <ul>
                            
                                                    <li><a href="https://www.facebook.com/ruan.zheng.7">Facebook</a></li>
                                                    <li><a href="http://zr1991.blogspot.com/">Google Blogger</a></li>
                                                    <li><a href="http://user.qzone.qq.com/8390905">Qzone</a></li>
                                                </ul>
                </div><!-- /.social -->
                </section><!-- /#extras -->

        <footer id="contentinfo" class="body">

        </footer><!-- /#contentinfo -->

<script type="text/javascript">
    var disqus_shortname = 'zruansblog';
    (function () {
        var s = document.createElement('script'); s.async = true;
        s.type = 'text/javascript';
        s.src = 'http://' + disqus_shortname + '.disqus.com/count.js';
        (document.getElementsByTagName('HEAD')[0] || document.getElementsByTagName('BODY')[0]).appendChild(s);
    }());
</script>
</body>
</html>